src/net/client/cache.rs

```diff
@@ -0,0 +1,1160 @@
+            return Ok(None);
+        }
+
+        let mut alt_key = key.clone();
+        alt_key.dnssec_ok = true;
+        alt_key.ad = true;
+        let opt_value = self.cache.get(&alt_key).await;
+        if let Some(value) = &opt_value {
+            match &value.response {
+                Err(_) => {
+                    // Just insert the error
+                    self.cache_insert(key.clone(), value.clone()).await;
+                }
+                Ok(msg) => {
+                    let msg = remove_dnssec(msg, key.ad)?;
+                    let value =
+                        Arc::new(Value::<C>::new_from_value_and_response(
+                            value.clone(),
+                            Ok(msg),
+                            &self.config,
+                        )?);
+                    self.cache_insert(key.clone(), value.clone()).await;
+                    return Ok(Some(value));
+                }
+            }
+        }
+        Ok(opt_value)
+    }
+
+    /// Try to find an cache entry for the key taking into account the
+    /// AD flag.
+    async fn cache_lookup_ad(
+        &self,
+        key: &Key,
+    ) -> Result<Option<Arc<Value<C>>>, Error> {
+        // For AD=1 we can only use responses to queries with AD set.
+        // For AD=0, first try with AD=0 and then try with AD=1. If
+        // AD=1 has an answer, clear the AD bit.
+        let opt_value = self.cache.get(key).await;
+        if opt_value.is_some() || key.ad {
+            return Ok(opt_value);
+        }
+        let mut alt_key = key.clone();
+        alt_key.ad = true;
+        let opt_value = self.cache.get(&alt_key).await;
+        if let Some(value) = &opt_value {
+            match &value.response {
+                Err(_) => {
+                    // Just insert the error
+                    self.cache_insert(key.clone(), value.clone()).await;
+                }
+                Ok(msg) => {
+                    if !msg.header().ad() {
+                        // AD is clear. Just insert this value.
+                        self.cache_insert(key.clone(), value.clone()).await;
+                        return Ok(Some(value.clone()));
+                    }
+
+                    let msg = clear_ad(msg)?;
+                    let value =
+                        Arc::new(Value::<C>::new_from_value_and_response(
+                            value.clone(),
+                            Ok(msg),
+                            &self.config,
+                        )?);
+                    self.cache_insert(key.clone(), value.clone()).await;
+                    return Ok(Some(value));
+                }
+            }
+        }
+        Ok(opt_value)
+    }
+
+    /// Insert new entry in the cache.
+    ///
+    /// Do not insert if the validity is zero.
+    /// Make sure to clear the AA flag.
+    async fn cache_insert(&self, key: Key, value: Arc<Value<C>>) {
+        if value.validity.is_zero() {
```

Comment:
Should this be even more aggressive and check for greater than some small threshold duration?

Reply:
No. is_zero is used as a signal that something should not be cached. That is consistent with the use of TTL value 0 in DNS. There is nothing wrong with caching something for a small period of time. Unless later that turns out to be wrong. But that seems unlikely because TTL is in seconds, and 1 second is an eternity on modern computers.